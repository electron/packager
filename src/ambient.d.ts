declare module 'get-package-info' {
  type Props = Array<string | string[]>

  export type GetPackageInfoResultSourceItem = {
    /** path to the package.json file */
    src: string

    /** property name */
    prop: string

    /** the `package.json` object */
    pkg: Record<string, unknown>
  }

  export type GetPackageInfoResult = {
    source: Record<string, GetPackageInfoResultSourceItem>
    values: Record<string, unknown>
  }

  function getPackageInfo (props: Props, dir: string): Promise<GetPackageInfoResult>

  export type GetPackageInfoError = Error & {
    missingProps: Props
    result: GetPackageInfoResult
  }

  export = getPackageInfo
}
