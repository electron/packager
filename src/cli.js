'use strict'

const { info, hostInfo, warning } = require('./common')
const fs = require('fs-extra')
const { initializeProxy } = require('@electron/get')
const packager = require('..')
const path = require('path')
const yargs = require('yargs-parser')

/* istanbul ignore next */
async function printUsageAndExit (isError) {
  const usage = (await fs.readFile(path.resolve(__dirname, '..', 'usage.txt'))).toString()
  const print = isError ? console.error : console.log
  print(usage)
  process.exit(isError ? 1 : 0)
}

module.exports = {
  parseArgs: function parseArgs (argv) {
    const args = yargs(argv, {
      boolean: [
        'all',
        'deref-symlinks',
        'download.rejectUnauthorized',
        'junk',
        'overwrite',
        'prune',
        'quiet'
      ],
      default: {
        'deref-symlinks': true,
        'download.rejectUnauthorized': true,
        junk: true,
        prune: true
      },
      string: [
        'electron-version',
        'out'
      ]
    })

    args.dir = args._[0]
    args.name = args._[1]

    const protocolSchemes = [].concat(args.protocol || [])
    const protocolNames = [].concat(args.protocolName || [])

    if (protocolSchemes && protocolNames && protocolNames.length === protocolSchemes.length) {
      args.protocols = protocolSchemes.map(function (scheme, i) {
        return { schemes: [scheme], name: protocolNames[i] }
      })
    }

    if (args.out === '') {
      warning('Specifying --out= without a value is the same as the default value')
      args.out = null
    }

    // Overrides for multi-typed arguments, because minimist doesn't support it

    // asar: `Object` or `true`
    if (args.asar === 'true' || args.asar instanceof Array) {
      warning('--asar does not take any arguments, it only has sub-properties (see --help)')
      args.asar = true
    }

    // osx-sign: `Object` or `true`
    if (args.osxSign === 'true') {
      warning('--osx-sign does not take any arguments, it only has sub-properties (see --help)')
      args.osxSign = true
    } else if (typeof args['osx-sign'] === 'object') {
      if (Array.isArray(args['osx-sign'])) {
        warning('Remove --osx-sign (the bare flag) from the command line, only specify sub-properties (see --help)')
      } else {
        // Keep kebab case of sub properties
        args.osxSign = args['osx-sign']
      }
    }

    if (args.osxNotarize) {
      let notarize = true
      if (typeof args.osxNotarize !== 'object' || Array.isArray(args.osxNotarize)) {
        warning('--osx-notarize does not take any arguments, it only has sub-properties (see --help)')
        notarize = false
      } else if (!args.osxSign) {
        warning('Notarization was enabled but macOS code signing was not, code signing is a requirement for notarization, notarize will not run')
        notarize = false
      }

      if (!notarize) {
        args.osxNotarize = null
      }
    }

    // tmpdir: `String` or `false`
    if (args.tmpdir === 'false') {
      warning('--tmpdir=false is deprecated, use --no-tmpdir instead')
      args.tmpdir = false
    }

    return args
  },
  run: /* istanbul ignore next */ async function run (argv) {
    const args = module.exports.parseArgs(argv)

    // temporary fix for https://github.com/nodejs/node/issues/6456
    for (const stdioWriter of [process.stdout, process.stderr]) {
      if (stdioWriter._handle && stdioWriter._handle.setBlocking) {
        stdioWriter._handle.setBlocking(true)
      }
    }

    if (args.help) {
      await printUsageAndExit(false)
    } else if (args.version) {
      if (typeof args.version !== 'boolean') {
        console.error('--version does not take an argument. Perhaps you meant --app-version or --electron-version?\n')
      }
      console.log(hostInfo())
      process.exit(0)
    } else if (!args.dir) {
      await printUsageAndExit(true)
    }

    initializeProxy()

    try {
      const appPaths = await packager(args)
      if (appPaths.length > 1) {
        info(`Wrote new apps to:\n${appPaths.join('\n')}`)
      } else if (appPaths.length === 1) {
        info('Wrote new app to', appPaths[0])
      }
    } catch (err) {
      if (err.message) {
        console.error(err.message)
      } else {
        console.error(err, err.stack)
      }
      process.exit(1)
    }
  }
}
