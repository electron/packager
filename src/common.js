'use strict'

const debug = require('debug')('electron-packager')
const filenamify = require('filenamify')
const fs = require('fs-extra')
const metadata = require('../package.json')
const os = require('os')
const path = require('path')

function sanitizeAppName (name) {
  return filenamify(name, { replacement: '-' })
}

function generateFinalBasename (opts) {
  return `${sanitizeAppName(opts.name)}-${opts.platform}-${opts.arch}`
}

function generateFinalPath (opts) {
  return path.join(opts.out || process.cwd(), generateFinalBasename(opts))
}

function info (message, quiet) {
  if (!quiet) {
    console.error(message)
  }
}

function warning (message, quiet) {
  if (!quiet) {
    console.warn(`WARNING: ${message}`)
  }
}

function subOptionWarning (properties, optionName, parameter, value, quiet) {
  if (Object.prototype.hasOwnProperty.call(properties, parameter)) {
    warning(`${optionName}.${parameter} will be inferred from the main options`, quiet)
  }
  properties[parameter] = value
}

function createAsarOpts (opts) {
  let asarOptions
  if (opts.asar === true) {
    asarOptions = {}
  } else if (typeof opts.asar === 'object') {
    asarOptions = opts.asar
  } else if (opts.asar === false || opts.asar === undefined) {
    return false
  } else {
    warning(`asar parameter set to an invalid value (${opts.asar}), ignoring and disabling asar`, opts.quiet)
    return false
  }

  return asarOptions
}

module.exports = {
  ensureArray: function ensureArray (value) {
    return Array.isArray(value) ? value : [value]
  },
  isPlatformMac: function isPlatformMac (platform) {
    return platform === 'darwin' || platform === 'mas'
  },

  createAsarOpts: createAsarOpts,

  deprecatedParameter: function deprecatedParameter (properties, oldName, newName, newCLIName, quiet) {
    if (Object.prototype.hasOwnProperty.call(properties, oldName)) {
      warning(`The ${oldName} parameter is deprecated, use ${newName} (or --${newCLIName} in the CLI) instead`, quiet)
      if (!Object.prototype.hasOwnProperty.call(properties, newName)) {
        properties[newName] = properties[oldName]
      }
      delete properties[oldName]
    }
  },
  subOptionWarning: subOptionWarning,

  baseTempDir: function baseTempDir (opts) {
    return path.join(opts.tmpdir || os.tmpdir(), 'electron-packager')
  },
  generateFinalBasename: generateFinalBasename,
  generateFinalPath: generateFinalPath,
  sanitizeAppName,
  /**
   * Convert slashes to UNIX-format separators.
   */
  normalizePath: function normalizePath (pathToNormalize) {
    return pathToNormalize.replace(/\\/g, '/')
  },
  /**
   * Validates that the application directory contains a package.json file, and that there exists an
   * appropriate main entry point file, per the rules of the "main" field in package.json.
   *
   * See: https://docs.npmjs.com/cli/v6/configuring-npm/package-json#main
   *
   * @param appDir - the directory specified by the user
   * @param bundledAppDir - the directory where the appDir is copied to in the bundled Electron app
   */
  validateElectronApp: async function validateElectronApp (appDir, bundledAppDir) {
    debug('Validating bundled Electron app')
    debug('Checking for a package.json file')

    const bundledPackageJSONPath = path.join(bundledAppDir, 'package.json')
    if (!(await fs.pathExists(bundledPackageJSONPath))) {
      const originalPackageJSONPath = path.join(appDir, 'package.json')
      throw new Error(`Application manifest was not found. Make sure "${originalPackageJSONPath}" exists and does not get ignored by your ignore option`)
    }

    debug('Checking for the main entry point file')
    const packageJSON = await fs.readJson(bundledPackageJSONPath)
    const mainScriptBasename = packageJSON.main || 'index.js'
    const mainScript = path.resolve(bundledAppDir, mainScriptBasename)
    if (!(await fs.pathExists(mainScript))) {
      const originalMainScript = path.join(appDir, mainScriptBasename)
      throw new Error(`The main entry point to your app was not found. Make sure "${originalMainScript}" exists and does not get ignored by your ignore option`)
    }

    debug('Validation complete')
  },

  hostInfo: function hostInfo () {
    return `Electron Packager ${metadata.version}\n` +
      `Node ${process.version}\n` +
      `Host Operating system: ${process.platform} ${os.release()} (${process.arch})`
  },
  info: info,
  warning: warning
}
