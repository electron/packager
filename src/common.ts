import filenamify from 'filenamify';
import fs from 'fs-extra';
import os from 'os';
import path from 'path';
import createDebug from 'debug';
import { ComboOptions, Options } from './types';
import { CreateOptions as AsarOptions } from '@electron/asar';

export const debug = createDebug('electron-packager');

export function sanitizeAppName(name: string) {
  return filenamify(name, { replacement: '-' });
}

export function generateFinalBasename(opts: Pick<ComboOptions, 'arch' | 'name' | 'platform'>) {
  return `${sanitizeAppName(opts.name!)}-${opts.platform}-${opts.arch}`;
}

export function generateFinalPath(opts: ComboOptions) {
  return path.join(opts.out || process.cwd(), generateFinalBasename(opts));
}

export function info(message: unknown, quiet?: boolean) {
  if (!quiet) {
    console.info(message);
  }
}

export function warning(message: unknown, quiet?: boolean) {
  if (!quiet) {
    console.warn(`WARNING: ${message}`);
  }
}

export function subOptionWarning(properties: Record<string, unknown>, optionName: string, parameter: string, value: unknown, quiet?: boolean) {
  if (Object.prototype.hasOwnProperty.call(properties, parameter)) {
    warning(`${optionName}.${parameter} will be inferred from the main options`, quiet);
  }
  properties[parameter] = value;
}

export function createAsarOpts(opts: ComboOptions): false | AsarOptions {
  let asarOptions;
  if (opts.asar === true) {
    asarOptions = {};
  } else if (typeof opts.asar === 'object') {
    asarOptions = opts.asar;
  } else if (opts.asar === false || opts.asar === undefined) {
    return false;
  } else {
    warning(`asar parameter set to an invalid value (${opts.asar}), ignoring and disabling asar`, opts.quiet);
    return false;
  }

  return asarOptions;
}

export function ensureArray<T>(value: T | T[]): T[] {
  return Array.isArray(value) ? value : [value];
}

export function isPlatformMac(platform: ComboOptions['platform']) {
  return platform === 'darwin' || platform === 'mas';
}

export function baseTempDir(opts: Options) {
  return path.join(opts.tmpdir || os.tmpdir(), 'electron-packager');
}

/**
 * Convert slashes to UNIX-format separators.
 */
export function normalizePath(pathToNormalize: string) {
  return pathToNormalize.replace(/\\/g, '/');
}

/**
 * Validates that the application directory contains a package.json file, and that there exists an
 * appropriate main entry point file, per the rules of the "main" field in package.json.
 *
 * See: https://docs.npmjs.com/cli/v6/configuring-npm/package-json#main
 *
 * @param appDir - the directory specified by the user
 * @param bundledAppDir - the directory where the appDir is copied to in the bundled Electron app
 */
export async function validateElectronApp(appDir: string, bundledAppDir: string) {
  debug('Validating bundled Electron app');
  debug('Checking for a package.json file');

  const bundledPackageJSONPath = path.join(bundledAppDir, 'package.json');
  if (!(await fs.pathExists(bundledPackageJSONPath))) {
    const originalPackageJSONPath = path.join(appDir, 'package.json');
    throw new Error(`Application manifest was not found. Make sure "${originalPackageJSONPath}" exists and does not get ignored by your ignore option`);
  }

  debug('Checking for the main entry point file');
  const packageJSON = await fs.readJson(bundledPackageJSONPath);
  const mainScriptBasename = packageJSON.main || 'index.js';
  const mainScript = path.resolve(bundledAppDir, mainScriptBasename);
  if (!(await fs.pathExists(mainScript))) {
    const originalMainScript = path.join(appDir, mainScriptBasename);
    throw new Error(`The main entry point to your app was not found. Make sure "${originalMainScript}" exists and does not get ignored by your ignore option`);
  }

  debug('Validation complete');
}

export function hostInfo() {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const metadata = require('../package.json');

  return `Electron Packager ${metadata.version}\n` +
    `Node ${process.version}\n` +
    `Host Operating system: ${process.platform} ${os.release()} (${process.arch})`;
}
