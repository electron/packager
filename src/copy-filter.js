'use strict'

const common = require('./common')
const debug = require('debug')('electron-packager')
const junk = require('junk')
const path = require('path')
const prune = require('./prune')
const targets = require('./targets')

const DEFAULT_IGNORES = [
  '/package-lock\\.json$',
  '/yarn\\.lock$',
  '/\\.git($|/)',
  '/node_modules/\\.bin($|/)',
  '\\.o(bj)?$',
  '/node_gyp_bins($|/)'
]

function populateIgnoredPaths (opts) {
  opts.originalIgnore = opts.ignore
  if (typeof (opts.ignore) !== 'function') {
    if (opts.ignore) {
      opts.ignore = common.ensureArray(opts.ignore).concat(DEFAULT_IGNORES)
    } else {
      opts.ignore = [].concat(DEFAULT_IGNORES)
    }
    if (process.platform === 'linux') {
      opts.ignore.push(common.baseTempDir(opts))
    }

    debug('Ignored path regular expressions:', opts.ignore)
  }
}

function generateIgnoredOutDirs (opts) {
  const normalizedOut = opts.out ? path.resolve(opts.out) : null
  const ignoredOutDirs = []
  if (normalizedOut === null || normalizedOut === process.cwd()) {
    for (const [platform, archs] of Object.entries(targets.officialPlatformArchCombos)) {
      for (const arch of archs) {
        const basenameOpts = {
          arch: arch,
          name: opts.name,
          platform: platform
        }
        ignoredOutDirs.push(path.join(process.cwd(), common.generateFinalBasename(basenameOpts)))
      }
    }
  } else {
    ignoredOutDirs.push(normalizedOut)
  }

  debug('Ignored paths based on the out param:', ignoredOutDirs)

  return ignoredOutDirs
}

function generateFilterFunction (ignore) {
  if (typeof (ignore) === 'function') {
    return file => !ignore(file)
  } else {
    const ignoredRegexes = common.ensureArray(ignore)

    return function filterByRegexes (file) {
      return !ignoredRegexes.some(regex => file.match(regex))
    }
  }
}

function userPathFilter (opts) {
  const filterFunc = generateFilterFunction(opts.ignore || [])
  const ignoredOutDirs = generateIgnoredOutDirs(opts)
  const pruner = opts.prune ? new prune.Pruner(opts.dir, opts.quiet) : null

  return async function filter (file) {
    const fullPath = path.resolve(file)

    if (ignoredOutDirs.includes(fullPath)) {
      return false
    }

    if (opts.junk !== false) { // defaults to true
      if (junk.is(path.basename(fullPath))) {
        return false
      }
    }

    let name = fullPath.split(path.resolve(opts.dir))[1]

    if (path.sep === '\\') {
      name = common.normalizePath(name)
    }

    if (pruner && name.startsWith('/node_modules/')) {
      if (await prune.isModule(file)) {
        return pruner.pruneModule(name)
      } else {
        return filterFunc(name)
      }
    }

    return filterFunc(name)
  }
}

module.exports = {
  populateIgnoredPaths,
  generateIgnoredOutDirs,
  userPathFilter
}
