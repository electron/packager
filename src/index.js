'use strict'

const common = require('./common')
const copyFilter = require('./copy-filter')
const debug = require('debug')('electron-packager')
const download = require('./download')
const fs = require('fs-extra')
const getMetadataFromPackageJSON = require('./infer')
const hooks = require('./hooks')
const path = require('path')
const targets = require('./targets')
const unzip = require('./unzip')
const { packageUniversalMac } = require('./universal')

function debugHostInfo () {
  debug(common.hostInfo())
}

class Packager {
  constructor (opts) {
    this.opts = opts
    this.tempBase = common.baseTempDir(opts)
    this.useTempDir = opts.tmpdir !== false
    this.canCreateSymlinks = undefined
  }

  async ensureTempDir () {
    if (this.useTempDir) {
      await fs.remove(this.tempBase)
    } else {
      return Promise.resolve()
    }
  }

  async testSymlink (comboOpts, zipPath) {
    const testPath = path.join(this.tempBase, `symlink-test-${comboOpts.platform}-${comboOpts.arch}`)
    const testFile = path.join(testPath, 'test')
    const testLink = path.join(testPath, 'testlink')

    try {
      await fs.outputFile(testFile, '')
      await fs.symlink(testFile, testLink)
      this.canCreateSymlinks = true
    } catch (e) {
      /* istanbul ignore next */
      this.canCreateSymlinks = false
    } finally {
      await fs.remove(testPath)
    }

    if (this.canCreateSymlinks) {
      return this.checkOverwrite(comboOpts, zipPath)
    }

    /* istanbul ignore next */
    return this.skipHostPlatformSansSymlinkSupport(comboOpts)
  }

  /* istanbul ignore next */
  skipHostPlatformSansSymlinkSupport (comboOpts) {
    common.info(`Cannot create symlinks (on Windows hosts, it requires admin privileges); skipping ${comboOpts.platform} platform`, this.opts.quiet)
    return Promise.resolve()
  }

  async overwriteAndCreateApp (outDir, comboOpts, zipPath) {
    debug(`Removing ${outDir} due to setting overwrite: true`)
    await fs.remove(outDir)
    return this.createApp(comboOpts, zipPath)
  }

  async extractElectronZip (comboOpts, zipPath, buildDir) {
    debug(`Extracting ${zipPath} to ${buildDir}`)
    await unzip(zipPath, buildDir)
    await hooks.promisifyHooks(this.opts.afterExtract, [buildDir, comboOpts.electronVersion, comboOpts.platform, comboOpts.arch])
  }

  buildDir (platform, arch) {
    let buildParentDir
    if (this.useTempDir) {
      buildParentDir = this.tempBase
    } else {
      buildParentDir = this.opts.out || process.cwd()
    }
    return path.resolve(buildParentDir, `${platform}-${arch}-template`)
  }

  async createApp (comboOpts, zipPath) {
    const buildDir = this.buildDir(comboOpts.platform, comboOpts.arch)
    common.info(`Packaging app for platform ${comboOpts.platform} ${comboOpts.arch} using electron v${comboOpts.electronVersion}`, this.opts.quiet)

    debug(`Creating ${buildDir}`)
    await fs.ensureDir(buildDir)
    await this.extractElectronZip(comboOpts, zipPath, buildDir)
    const os = require(targets.osModules[comboOpts.platform])
    const app = new os.App(comboOpts, buildDir)
    return app.create()
  }

  async checkOverwrite (comboOpts, zipPath) {
    const finalPath = common.generateFinalPath(comboOpts)
    if (await fs.pathExists(finalPath)) {
      if (this.opts.overwrite) {
        return this.overwriteAndCreateApp(finalPath, comboOpts, zipPath)
      } else {
        common.info(`Skipping ${comboOpts.platform} ${comboOpts.arch} (output dir already exists, use --overwrite to force)`, this.opts.quiet)
        return true
      }
    } else {
      return this.createApp(comboOpts, zipPath)
    }
  }

  async getElectronZipPath (downloadOpts) {
    if (this.opts.electronZipDir) {
      if (await fs.pathExists(this.opts.electronZipDir)) {
        const zipPath = path.resolve(
          this.opts.electronZipDir,
          `electron-v${downloadOpts.version}-${downloadOpts.platform}-${downloadOpts.arch}.zip`
        )
        if (!await fs.pathExists(zipPath)) {
          throw new Error(`The specified Electron ZIP file does not exist: ${zipPath}`)
        }

        return zipPath
      }

      throw new Error(`The specified Electron ZIP directory does not exist: ${this.opts.electronZipDir}`)
    } else {
      return download.downloadElectronZip(downloadOpts)
    }
  }

  async packageForPlatformAndArchWithOpts (comboOpts, downloadOpts) {
    const zipPath = await this.getElectronZipPath(downloadOpts)

    if (!this.useTempDir) {
      return this.createApp(comboOpts, zipPath)
    }

    if (common.isPlatformMac(comboOpts.platform)) {
      /* istanbul ignore else */
      if (this.canCreateSymlinks === undefined) {
        return this.testSymlink(comboOpts, zipPath)
      } else if (!this.canCreateSymlinks) {
        return this.skipHostPlatformSansSymlinkSupport(comboOpts)
      }
    }

    return this.checkOverwrite(comboOpts, zipPath)
  }

  async packageForPlatformAndArch (downloadOpts) {
    // Create delegated options object with specific platform and arch, for output directory naming
    const comboOpts = {
      ...this.opts,
      arch: downloadOpts.arch,
      platform: downloadOpts.platform,
      electronVersion: downloadOpts.version
    }

    if (common.isPlatformMac(comboOpts.platform) && comboOpts.arch === 'universal') {
      return packageUniversalMac(this.packageForPlatformAndArchWithOpts.bind(this), this.buildDir(comboOpts.platform, comboOpts.arch), comboOpts, downloadOpts, this.tempBase)
    }

    return this.packageForPlatformAndArchWithOpts(comboOpts, downloadOpts)
  }
}

async function packageAllSpecifiedCombos (opts, archs, platforms) {
  const packager = new Packager(opts)
  await packager.ensureTempDir()
  return Promise.all(download.createDownloadCombos(opts, platforms, archs).map(
    downloadOpts => packager.packageForPlatformAndArch(downloadOpts)
  ))
}

module.exports = async function packager (opts) {
  debugHostInfo()
  if (debug.enabled) debug(`Packager Options: ${JSON.stringify(opts)}`)

  const archs = targets.validateListFromOptions(opts, 'arch')
  const platforms = targets.validateListFromOptions(opts, 'platform')
  if (!Array.isArray(archs)) return Promise.reject(archs)
  if (!Array.isArray(platforms)) return Promise.reject(platforms)

  debug(`Target Platforms: ${platforms.join(', ')}`)
  debug(`Target Architectures: ${archs.join(', ')}`)

  const packageJSONDir = path.resolve(process.cwd(), opts.dir) || process.cwd()

  await getMetadataFromPackageJSON(platforms, opts, packageJSONDir)
  if (opts.name.endsWith(' Helper')) {
    throw new Error('Application names cannot end in " Helper" due to limitations on macOS')
  }

  debug(`Application name: ${opts.name}`)
  debug(`Target Electron version: ${opts.electronVersion}`)

  copyFilter.populateIgnoredPaths(opts)

  const appPaths = await packageAllSpecifiedCombos(opts, archs, platforms)
  // Remove falsy entries (e.g. skipped platforms)
  return appPaths.filter(appPath => appPath)
}
