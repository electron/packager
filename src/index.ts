import { serialHooks } from './hooks';
import { packager } from './packager';
import { allOfficialArchsForPlatformAndVersion } from './targets';

export default packager;

export { allOfficialArchsForPlatformAndVersion, packager, serialHooks };

export * from './types';
