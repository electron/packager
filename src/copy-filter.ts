import { baseTempDir, debug, ensureArray, generateFinalBasename, normalizePath } from './common';
import junk from 'junk';
import path from 'path';
import { isModule, Pruner } from './prune';
import { officialPlatformArchCombos } from './targets';
import { ComboOptions, Options } from './types';
import { CopyFilterAsync } from 'fs-extra';

const DEFAULT_IGNORES = [
  '/package-lock\\.json$',
  '/yarn\\.lock$',
  '/\\.git($|/)',
  '/node_modules/\\.bin($|/)',
  '\\.o(bj)?$',
  '/node_gyp_bins($|/)',
];

export function populateIgnoredPaths(opts: Options) {
  (opts as Options & { originalIgnore: Options['ignore'] }).originalIgnore = opts.ignore;

  if (typeof (opts.ignore) !== 'function') {
    if (opts.ignore) {
      opts.ignore = [...ensureArray(opts.ignore), ...DEFAULT_IGNORES];
    } else {
      opts.ignore = [...DEFAULT_IGNORES];
    }
    if (process.platform === 'linux') {
      opts.ignore.push(baseTempDir(opts));
    }

    debug('Ignored path regular expressions:', opts.ignore);
  }
}

export function generateIgnoredOutDirs(opts: ComboOptions): string[] {
  const normalizedOut = opts.out ? path.resolve(opts.out) : null;
  const ignoredOutDirs: string[] = [];

  if (normalizedOut === null || normalizedOut === process.cwd()) {
    for (const [platform, archs] of Object.entries(officialPlatformArchCombos)) {
      for (const arch of archs) {
        const basenameOpts = {
          arch: arch,
          name: opts.name,
          platform: platform,
        };
        ignoredOutDirs.push(path.join(process.cwd(), generateFinalBasename(basenameOpts)));
      }
    }
  } else {
    ignoredOutDirs.push(normalizedOut);
  }

  debug('Ignored paths based on the out param:', ignoredOutDirs);

  return ignoredOutDirs;
}

function generateFilterFunction(ignore: Exclude<ComboOptions['ignore'], undefined>): (file: string) => boolean {
  if (typeof (ignore) === 'function') {
    return file => !ignore(file);
  } else {
    const ignoredRegexes = ensureArray(ignore);

    return function filterByRegexes(file) {
      return !ignoredRegexes.some(regex => file.match(regex));
    };
  }
}

export function userPathFilter(opts: ComboOptions): CopyFilterAsync {
  const filterFunc = generateFilterFunction(opts.ignore || []);
  const ignoredOutDirs = generateIgnoredOutDirs(opts);
  const pruner = opts.prune ? new Pruner(opts.dir, Boolean(opts.quiet)) : null;

  return async function filter(file) {
    const fullPath = path.resolve(file);

    if (ignoredOutDirs.includes(fullPath)) {
      return false;
    }

    if (opts.junk !== false) { // defaults to true
      if (junk.is(path.basename(fullPath))) {
        return false;
      }
    }

    let name = fullPath.split(path.resolve(opts.dir))[1];

    if (path.sep === '\\') {
      name = normalizePath(name);
    }

    if (pruner && name.startsWith('/node_modules/')) {
      if (await isModule(file)) {
        return pruner.pruneModule(name);
      } else {
        return filterFunc(name);
      }
    }

    return filterFunc(name);
  } as CopyFilterAsync;
}
