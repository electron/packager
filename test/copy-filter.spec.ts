import {
  generateIgnoredOutDirs,
  populateIgnoredPaths,
  userPathFilter,
} from '../src/copy-filter';
import fs from 'fs-extra';
import os from 'node:os';
import path from 'node:path';
import { describe, it, expect } from 'vitest';
import { ComboOptions, Options } from '../src';

describe('populateIgnoredPaths', () => {
  let tempDir: string;
  beforeEach(async () => {
    tempDir = await fs.mkdtemp(
      path.join(os.tmpdir(), 'electron-packager-test-'),
    );
  });

  it('ignores the generated temporary directory only on Linux', () => {
    const tmpdir = '/foo/bar';
    const expected = path.join(tmpdir, 'electron-packager');
    const opts = { tmpdir } as Options;

    populateIgnoredPaths(opts);

    if (process.platform === 'linux') {
      expect(opts.ignore).toContain(expected);
    } else {
      expect(opts.ignore).not.toContain(expected);
    }
  });

  it('ignores certain files by default', async () => {
    const opts = {
      name: 'test',
      dir: path.join(__dirname, 'fixtures', 'basic'),
    } as Options;

    populateIgnoredPaths(opts);
    const targetDir = path.join(tempDir, 'result');
    await fs.copy(opts.dir, targetDir, {
      dereference: false,
      filter: userPathFilter(opts as ComboOptions),
    });

    expect(fs.existsSync(path.join(targetDir, 'node_gyp_bins'))).toBe(false);
    expect(fs.existsSync(path.join(targetDir, 'ignore.o'))).toBe(false);
    expect(fs.existsSync(path.join(targetDir, 'ignore.obj'))).toBe(false);
    expect(fs.existsSync(path.join(targetDir, 'package-lock.json'))).toBe(
      false,
    );
    expect(fs.existsSync(path.join(targetDir, 'pnpm-lock.yaml'))).toBe(false);
    expect(fs.existsSync(path.join(targetDir, 'yarn.lock'))).toBe(false);
    expect(fs.existsSync(path.join(targetDir, 'Thumbs.db'))).toBe(false);

    // Random txt files should not be ignored
    expect(fs.existsSync(path.join(targetDir, 'ignorethis.txt'))).toBe(true);
  });
});

describe('userPathFilter', () => {
  let tempDir: string;
  beforeEach(async () => {
    tempDir = await fs.mkdtemp(
      path.join(os.tmpdir(), 'electron-packager-test-'),
    );
  });
  it.each([
    ['string', 'ignorethis'],
    ['string in array', ['ignorethis']],
    ['RegExp', /ignorethis/],
    ['Function', (file: string) => file.match(/ignorethis/)],
  ])('ignores %s that match the ignore pattern', async (_, ignore) => {
    const opts = {
      name: 'test',
      dir: path.join(__dirname, 'fixtures', 'basic'),
      ignore: ignore,
    } as Options;

    populateIgnoredPaths(opts);
    const targetDir = path.join(tempDir, 'result');
    await fs.copy(opts.dir, targetDir, {
      dereference: false,
      filter: userPathFilter(opts as ComboOptions),
    });

    expect(fs.existsSync(path.join(targetDir, 'ignorethis.txt'))).toBe(false);
    expect(fs.existsSync(path.join(targetDir, 'ignore', 'this.txt'))).toBe(
      true,
    );
  });

  it('parses slashes in the ignore pattern', async () => {
    const opts = {
      name: 'test',
      dir: path.join(__dirname, 'fixtures', 'basic'),
      ignore: ['ignore/this'],
    } as Options;

    populateIgnoredPaths(opts);
    const targetDir = path.join(tempDir, 'result');
    await fs.copy(opts.dir, targetDir, {
      dereference: false,
      filter: userPathFilter(opts as ComboOptions),
    });

    expect(fs.existsSync(path.join(targetDir, 'ignorethis.txt'))).toBe(true);
    expect(fs.existsSync(path.join(targetDir, 'ignore', 'this.txt'))).toBe(
      false,
    );
  });

  it('only ignores files within the app directory', async () => {
    const opts = {
      name: 'test',
      dir: path.join(__dirname, 'fixtures', 'basic'),
      // we're testing to see that not all files are ignored from the source dir if a parent directory is in the ignore list
      ignore: [__dirname],
    } as Options;

    populateIgnoredPaths(opts);

    const targetDir = path.join(tempDir, 'result');
    await fs.copy(opts.dir, targetDir, {
      dereference: false,
      filter: userPathFilter(opts as ComboOptions),
    });

    expect(fs.existsSync(path.join(targetDir, 'package.json'))).toBe(true);
  });

  it.each([
    ['junk: true', true],
    ['junk: false', false],
  ])('respects %s', async (_, junk) => {
    const opts = {
      name: 'test',
      dir: path.join(__dirname, 'fixtures', 'ignore-junk'),
      junk,
    } as Options;

    populateIgnoredPaths(opts);

    const targetDir = path.join(tempDir, 'result');
    await fs.copy(opts.dir, targetDir, {
      dereference: false,
      filter: userPathFilter(opts as ComboOptions),
    });

    expect(fs.existsSync(path.join(targetDir, 'subfolder', 'Thumbs.db'))).toBe(
      !junk,
    );
  });
});

describe('generateIgnoredOutDirs', () => {
  it('ignores all possible platform/arch permutations', () => {
    const ignores = generateIgnoredOutDirs({ name: 'test' } as ComboOptions);
    expect(ignores.length).toBe(14);
  });
});
