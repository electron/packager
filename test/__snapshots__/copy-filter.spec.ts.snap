// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateIgnoredOutDirs > ignores all possible platform/arch permutations 1`] = `
[
  "test-darwin-x64",
  "test-darwin-arm64",
  "test-darwin-universal",
  "test-linux-ia32",
  "test-linux-x64",
  "test-linux-armv7l",
  "test-linux-arm64",
  "test-linux-mips64el",
  "test-mas-x64",
  "test-mas-arm64",
  "test-mas-universal",
  "test-win32-ia32",
  "test-win32-x64",
  "test-win32-arm64",
]
`;
